Initial Azure.Core Perf work with two scenarios.

sa7936 opened the pull request:
-Worked on making the Benchmark.Nuget and Benchmark.Local projects with scenarios for the default pipeline and event source first.

-Currently compares the nuget vs local version of azure.core and runs powershell script to see if the mean runtimes are within 1 second of each other to pass.

 -Started work on integrating the perf tests into the pipeline (incomplete).

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

Commit:
-Worked on  making the Benhcmark.Nuget and Benchmark.Local projects with scenarios for the default pipeline and event source first.
-Currently compares the nuget vs local version of azure.core and runs powershell script to see if the mean runtimes are within 1 second of each other to pass.
-Started work on integrating the perf tests into the pipeline (incomplete).

Commit:
Removed unnecessary program.cs files and changed Benchmarks.local and benchmarks.nuget to class libraries.

Commit:
Trying to add the perf tests to the CI pipeline.

Commit:
Moved the coreperf job to be outside the else of the servicedirectory conditional.

Commit:
Moving the run-and-compare script to the eng/scripts folder.

Commit:
Fixing cross-platform error from core-perf script

Commit:
Moving the benchmarks.yml to be a separate job and pipeline

Commit:
Moving the job back into ci and setting it up in the pullrequest.yml

Commit:
Adding it to the build step in the pipeline

Commit:
Removing things from the pullrequest and archetype-sdk-client yml files.

m-redding commented on sdk/core/Azure.Core/Benchmarks.Local/Benchmarks.Local.csproj:
```diff
@@ -0,0 +1,19 @@
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+  <PropertyGroup>
+    <SignAssembly>true</SignAssembly>
```
Curious - why do you need this?

sa7936 replied:
EventSourceBenchmark used an unsafe method in testing RequestNew. The SignAssembly was a fix for a CS8002 error (referenced assmbly does not have a strong name"

m-redding replied:
Hmm I don't think we should need this. You may need to add a Directory.Build.props, something like this - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/integration/Directory.Build.props in the perf/ directory, and then add something like this to each project directory https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/integration/WebApp/Directory.Build.props

Commit:
Mocking the Http calls in the pipelinescenarios.

sa7936 commented on sdk/core/Azure.Core/Benchmarks.Nuget/PipelineScenario.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+            var message = _pipeline.CreateMessage();
+            message.Request.Uri.Reset(new Uri("https://www.example.com"));
+            await _pipeline.SendAsync(message, CancellationToken.None);
```
I put back the mocking of transport calls

Commit:
Removed the .snk files and added Drectory.Build.props files in replacement.

Commit:
Moved the benchmarks projects into the same folder with azure.core.perf and moved the nested classes into new files.

m-redding commented on sdk/core/Azure.Core/perf/Benchmarks.Local/PipelineScenario.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <returns>The HTTP response.</returns>
+        public async Task<Azure.Response> SendAsync()
+        {
+            var message = _pipeline.CreateMessage();
```
I think for the default case this works but it could be interesting to look at also having a benchmark that creates the messages in the setup and then benchmarks the sends only just to measure the send performance.

m-redding commented on sdk/core/Azure.Core/perf/Benchmarks.Nuget/BenchmarkClientOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        /// <summary>
+        /// The versions of the service supported by the client.
+        /// </summary>
+        public enum ServiceVersion
```
Same as other file

sa7936 committed:
Merge branch 'Azure:main' into t-akpoyibos/benchmark-work

Commit:
-Made a pipelinescenario that tests only the send functionality.
-Changed the compare script to use a combination of a percentage difference and and absolute threshold of 1000 ns (because sometimes there is a huge percwntage difference but the speed is in ns so the difference doesn't cleanly translate to slow). Still woking on refining script.
- Started work to make scenario for dynamicobjectbenchmark.

Commit:
--- Removed pipeline parts from the yml files and restored them to what they originally were.
--- Worked on the DynamicObjectBenchmarks and edited the scripts to run them.
---Modified azure.core.tests classes to be public

Commit:
Removed random white spaces in the yml files.

m-redding committed:
pass on different layout

m-redding committed:
tweak script

m-redding committed:
add shared source

sa7936 committed:
Merge branch 'Azure:main' into t-akpoyibos/benchmark-work

Commit:
Merge branch 't-akpoyibos/benchmark-work' of https://github.com/sa7936/azure-sdk-for-net into t-akpoyibos/benchmark-work

Commit:
Updated the files and compare script to save the local and nuget results in separate folders for easy comparison.

m-redding commented on sdk/core/Azure.Core/perf/EventSourceBenchmark.cs:
```diff
@@ -56,104 +55,35 @@ public void CleanUp()
         }
 
-        [Benchmark(Description = "New implementation, no reformatting")]
+        [Benchmark]
```
It seems like since the sanitizing/formatting headers was moved to set up, this "after formatting" benchmark might not be needed?

m-redding commented on sdk/core/Azure.Core/perf/CustomEventSource.cs:
```diff
@@ -0,0 +1,83 @@
+/// <summary>
+/// Custom event source for Azure Core logging.
+/// </summary>
+public class CustomEventSource : EventSource
```
It looks like this used to inherit from AzureEventSource and now it's EventSource, was this intentional to test something specific?

m-redding commented on sdk/core/Azure.Core/perf/PipelineBenchmark.cs:
```diff
@@ -10,79 +10,73 @@
-    }
+            _messageIndex = (_messageIndex + 1) % _messages.Length;
+
+            await _pipeline.SendAsync(message, CancellationToken.None).ConfigureAwait(false);
```
Nice I like the changes to this one!!

m-redding commented on sdk/core/Azure.Core/perf/PipelineDefaultBenchmark.cs:
```diff
@@ -0,0 +1,47 @@
+    [SimpleJob(RuntimeMoniker.Net80)]
+
+    [MemoryDiagnoser]
+    public class PipelineDefaultBenchmark
```
nit - maybe rename this to CreateAndSendWithDefaultPipeline or something similar? Just to more clearly differentiate it from PipelineBenchmark

Commit:
--- Renamed pipeline benchmanmrks to be more expressive of what they do.
--- Fixed the benchmarks in EventSourceBecnhmarks to test old_implementation with formatting, new_implementation with formatting (of Uri) and new_implementation after formatting.
--- Fixed the CustomEventSource to inherit from the AzureEventSource
---Restored Azure.Core tests to being internal
--- Removed properties no longer needed from AssemblyInfo.

m-redding commented on sdk/core/Azure.Core/Azure.Core.sln:
```diff
@@ -69,6 +67,12 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.ClientModel.SourceGe
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.ClientModel.SourceGeneration.Unit.Tests", "..\System.ClientModel\tests\gen.unit\System.ClientModel.SourceGeneration.Unit.Tests.csproj", "{6ED66F0E-C7FA-3F0E-38C8-D576D48C8A58}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Benchmarks.Local", "perf\Benchmarks.Local\Benchmarks.Local.csproj", "{9C308AA1-7177-8B0E-6D84-26E9169EBADA}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Benchmarks.Nuget", "perf\Benchmarks.Nuget\Benchmarks.Nuget.csproj", "{5CF26598-DC64-448B-6F81-D3CE1185FD78}"
+EndProject
```
```suggestion
```
To fix pipeline error

m-redding commented on sdk/core/Azure.Core/Azure.Core.sln:
```diff
@@ -179,6 +179,18 @@ Global
 		{6ED66F0E-C7FA-3F0E-38C8-D576D48C8A58}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{6ED66F0E-C7FA-3F0E-38C8-D576D48C8A58}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{6ED66F0E-C7FA-3F0E-38C8-D576D48C8A58}.Release|Any CPU.Build.0 = Release|Any CPU
+		{9C308AA1-7177-8B0E-6D84-26E9169EBADA}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{9C308AA1-7177-8B0E-6D84-26E9169EBADA}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{9C308AA1-7177-8B0E-6D84-26E9169EBADA}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{9C308AA1-7177-8B0E-6D84-26E9169EBADA}.Release|Any CPU.Build.0 = Release|Any CPU
+		{5CF26598-DC64-448B-6F81-D3CE1185FD78}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{5CF26598-DC64-448B-6F81-D3CE1185FD78}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{5CF26598-DC64-448B-6F81-D3CE1185FD78}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{5CF26598-DC64-448B-6F81-D3CE1185FD78}.Release|Any CPU.Build.0 = Release|Any CPU
```
```suggestion
```
same as above

m-redding commented on sdk/core/Azure.Core/perf/DynamicObjectBenchmark.cs:
```diff
@@ -12,19 +12,24 @@
         private ModelWithBinaryData _modelWithBinaryData;
         private ModelWithObject _modelWithObject;
 
+        private static readonly string FileName = Path.Combine(
```
```suggestion
        private static readonly string _fileName = Path.Combine(
```
nit - in the repo we usually try to have private fields use this naming style

m-redding commented on sdk/core/Azure.Core/perf/DynamicObjectBenchmark.cs:
```diff
@@ -12,19 +12,24 @@
         [GlobalSetup]
         public void SetUp()
         {
+            _fileName = FileName;
```
```suggestion
```

m-redding commented on sdk/core/Azure.Core/perf/EventSourceBenchmark.cs:
```diff
@@ -56,16 +56,17 @@ public void CleanUp()
         public void OldImplementation_CallFormatting()
         {
-            _eventSource.RequestOld(Sanitizer.SanitizeUrl(_uri.ToString()), _iteration++, _iteration, FormatHeaders());
+            _eventSource.RequestOld(Sanitizer.SanitizeUrl(_uri.ToString()), _iteration++, _iteration, _headersBytes);
```
does the header formatting need to be kept too in addition to the sanitization call?

m-redding replied:
If there wasn't a specific reason that you wanted to test the scenario without it, I think it would make sense to leave it in,

Commit:
---Removed Benchmark.Local and Nuget content from the .sln.
---Renamed private method to go with standards

Commit:
Editted the old implementation, to have formatting headers as part of the call.

Commit:
Copilot caught some typos and use of null-forgiving operator that was fixed.

m-redding commented on sdk/core/Azure.Core/perf/Azure.Core.Perf.csproj:
```diff
@@ -3,16 +3,17 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <TargetFrameworks>net462;net8.0;net9.0</TargetFrameworks>
```
```suggestion
```
There's some sort of weird pipeline error
```
/mnt/vss/_work/1/s/sdk/core/Azure.Core/perf/Azure.Core.Perf.csproj : error NU1201: Project Azure.Test.Perf is not compatible with net462 (.NETFramework,Version=v4.6.2). Project Azure.Test.Perf supports: [/mnt/vss/_work/1/s/eng/service.proj]
/mnt/vss/_work/1/s/sdk/core/Azure.Core/perf/Azure.Core.Perf.csproj : error NU1201:   - net8.0 (.NETCoreApp,Version=v8.0) [/mnt/vss/_work/1/s/eng/service.proj]
/mnt/vss/_work/1/s/sdk/core/Azure.Core/perf/Azure.Core.Perf.csproj : error NU1201:   - net9.0 (.NETCoreApp,Version=v9.0) [/mnt/vss/_work/1/s/eng/service.proj]
```
I'm wondering if this line needs to be removed.

sa7936 replied:
Yup, I added it when working with the other projects, but it's not needed anymore.

Commit:
Removed the targetframeworks property (not needed anymore).

Commit:
Removed targetframeworks change from Sys.clientmodel.tests.internal.perf.csproj

m-redding commented:
Failures are from Azure.Storage.Blobs - /mnt/vss/_work/1/s/sdk/storage/Azure.Storage.Blobs/tests/BlobClientTests.cs(69,82): error CS0618: 'SharedTokenCacheCredential' is obsolete: 'This credential is deprecated. Consider using other dev tool credentials, such as VisualStudioCredential.' [/mnt/vss/_work/1/s/sdk/storage/Azure.Storage.Blobs/tests/Azure.Storage.Blobs.Tests.csproj::TargetFramework=net9.0]

And Azure.Developer.MicrosoftPlaywrightTest.TestLogger 
/mnt/vss/_work/1/s/sdk/playwrighttesting/Azure.Developer.MicrosoftPlaywrightTesting.TestLogger/src/PlaywrightServiceOptions.cs(86,24): error CS0618: 'SharedTokenCacheCredential' is obsolete: 'This credential is deprecated. Consider using other dev tool credentials, such as VisualStudioCredential.' [/mnt/vss/_work/1/s/sdk/playwrighttesting/Azure.Developer.MicrosoftPlaywrightTesting.TestLogger/src/Azure.Developer.MicrosoftPlaywrightTesting.TestLogger.csproj::TargetFramework=net8.0]

Both of which are unrelated

m-redding commented:
/check-enforcer override
